File: backend/lumen_schemas/workspace_member.py
"""M38.2.x v2: workspace member management schemas.

Request / response shapes for the 5 member-management endpoints:

- ``GET    /api/v1/workspaces/{id}/members``  → ``WorkspaceMemberListResponse``
- ``POST   /api/v1/workspaces/{id}/members``  body: ``WorkspaceMemberInvite``
- ``PUT    /api/v1/workspaces/{id}/members/{user_id}``  body: ``WorkspaceMemberUpdate``
- ``DELETE /api/v1/workspaces/{id}/members/{user_id}``
- ``POST   /api/v1/workspaces/{id}/transfer-ownership``  body: ``WorkspaceTransferOwnership``

The 17 permission tokens live in
``lumen_services.permission_service._PERM_IMPLIES`` — this file only
declares Pydantic containers, never the permission enum, so adding a
new permission does NOT require a schema migration.

Spec: ``docs-internal/superpowers/specs/2026-08-27-workspace-rbac.md`` § 5.1.
"""
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class WorkspaceMemberInvite(BaseModel):
    """Body of ``POST /api/v1/workspaces/{id}/members``.

    ``permissions`` is the initial grant set; duplicate or invalid
    tokens → 400. Re-granting a permission the user already has is a
    no-op (the INSERT 409s are caught and turned into
    ``already_granted`` 409 responses by the service).
    """

    user_id: int = Field(..., description="目标 user.id;必须与 workspace 同 tenant")
    permissions: List[str] = Field(
        default_factory=list,
        description="授予的 permission token 列表(17 项任选)",
    )


class WorkspaceMemberRead(BaseModel):
    """Single member row in the response of ``GET /members``.

    Permission set is denormalized into a sorted list so the frontend
    can render a permission matrix without a second round-trip.
    """

    user_id: int
    username: Optional[str] = None
    permissions: List[str] = Field(default_factory=list)
    granted_at: Optional[datetime] = None
    # Marked True when this user is also the workspace owner — UI uses
    # this to render the owner pill and disable self-removal.
    is_owner: bool = False


class WorkspaceMemberListResponse(BaseModel):
    """Response of ``GET /api/v1/workspaces/{id}/members``.

    Includes the workspace ``id`` (echoed) so the frontend doesn't
    have to thread it through separately.
    """

    workspace_id: int
    items: List[WorkspaceMemberRead] = Field(default_factory=list)


class WorkspaceMemberUpdate(BaseModel):
    """Body of ``PUT /api/v1/workspaces/{id}/members/{user_id}``.

    Whole-set overwrite: rows whose ``permission`` is NOT in the new
    list are deleted; rows whose ``permission`` is in the new list
    but missing are added. This is the "set the user set" pattern —
    explicit deletion of the resource is a separate
    ``DELETE /members/{user_id}`` call.

    An empty list removes ALL grants (effectively same as DELETE on
    the membership row, except DELETE also wipes the audit trail).
    """

    permissions: List[str] = Field(
        default_factory=list,
        description="整组覆盖;在列表外的会被 DELETE,在列表内缺失会被 INSERT",
    )


class WorkspaceTransferOwnership(BaseModel):
    """Body of ``POST /api/v1/workspaces/{id}/transfer-ownership``.

    The new owner auto-grants ALL permissions via the ``owner``
    bypass in ``PermissionService.check`` — no row insertion is
    performed. The old owner keeps their current grants (typically
    full perm if they were already an admin).
    """

    new_owner_id: int = Field(..., description="新 owner 的 users.id;必须同 tenant")

File: frontend/types/workspaceMember.ts
// M38.2.x v2: Workspace member / RBAC permission types.
//
// Spec: docs-internal/superpowers/specs/2026-08-27-workspace-rbac.md §5.1.
// 19 个 permission token 跨 4 资源轴(workspace / kb / folder / document),
// 服务端 PermissionService 内部展开 implication 链(例 kb.update 自动含
// kb.read + document.read),前端只关心原始 19 项。

export type WorkspacePermission =
  | "workspace.read"
  | "workspace.update"
  | "workspace.delete"
  | "workspace.transfer_ownership"
  | "workspace.manage_members"
  | "kb.read"
  | "kb.create"
  | "kb.update"
  | "kb.delete"
  | "folder.read"
  | "folder.create"
  | "folder.update"
  | "folder.delete"
  | "folder.restore"
  | "document.read"
  | "document.create"
  | "document.update"
  | "document.delete"
  | "document.move";

export const ALL_PERMISSIONS: WorkspacePermission[] = [
  "workspace.read",
  "workspace.update",
  "workspace.delete",
  "workspace.transfer_ownership",
  "workspace.manage_members",
  "kb.read",
  "kb.create",
  "kb.update",
  "kb.delete",
  "folder.read",
  "folder.create",
  "folder.update",
  "folder.delete",
  "folder.restore",
  "document.read",
  "document.create",
  "document.update",
  "document.delete",
  "document.move",
];

/** 只读 4 项:用于「只读」快捷批量授权。 */
export const READ_ONLY_PERMISSIONS: WorkspacePermission[] = [
  "workspace.read",
  "kb.read",
  "folder.read",
  "document.read",
];

/** 写权限 = read + create/update/move(不含 delete / manage)。 */
export const WRITE_PERMISSIONS: WorkspacePermission[] = [
  "workspace.read",
  "workspace.update",
  "kb.read",
  "kb.create",
  "kb.update",
  "folder.read",
  "folder.create",
  "folder.update",
  "document.read",
  "document.create",
  "document.update",
  "document.move",
];

/** 单条 grant 行:某个 user 在某个 workspace 上持有的 permission。 */
export interface WorkspaceMemberPermission {
  workspace_id: number;
  user_id: number;
  permission: WorkspacePermission;
  created_at?: string;
}

/** 列出 workspace 全部成员(按 user 聚合)。 */
export interface WorkspaceMember {
  user_id: number;
  username: string;
  email?: string | null;
  full_name?: string | null;
  /** 该 user 在该 workspace 上的全部 grant 行(可能含重复 owner_id 不算)。 */
  permissions: WorkspacePermission[];
  /** 由后端 join 计算:is_owner = workspace.owner_id == user_id。 */
  is_owner: boolean;
  created_at?: string;
}

export interface MemberInvitePayload {
  user_id: number;
  /** 整组覆盖:邀请时一次性给一组权限,后续再调整。 */
  permissions: WorkspacePermission[];
}

export interface MemberUpdatePayload {
  /** 整组覆盖(POST 同形):用新集合替换旧集合,空集 = 撤销全部。 */
  permissions: WorkspacePermission[];
}

export interface TransferOwnershipPayload {
  new_owner_id: number;
}

export interface WorkspaceMembersResponse {
  members: WorkspaceMember[];
  total: number;
}

export interface WorkspaceMyPermissionsResponse {
  /** 当前 user 在各 workspace 上的 effective permission set。 */
  workspaces: Array<{
    workspace_id: number;
    permissions: WorkspacePermission[];
    is_owner: boolean;
  }>;
}